src/sentry/lang/native/cfi.py

```diff
@@ -0,0 +1,267 @@
+        # interface. Note that this is done with the same function as the
+        # initial ingestion to avoid normalization conflicts.
+        frames = frames_from_minidump_thread(thread)
+
+        # Filter out stack traces that did not improve during reprocessing. For
+        # these cases we only store a marker. This also prevents us from
+        # destroying absolute addresses when restoring from the cache. Stack
+        # traces containing CFI frames are mapped to their modules and stored.
+        if any(frame['trust'] in CFI_TRUSTS for frame in frames):
+            self.resolved_frames = [(self.modules.find_object(f['instruction_addr']), f)
+                                    for f in frames]
+        else:
+            self.resolved_frames = NO_CFI_PLACEHOLDER
+
+    def apply_to_event(self):
+        """Writes the loaded stack trace back to the event's payload. Returns
+        ``True`` if the payload was changed, otherwise ``False``."""
+        if self.resolved_frames is None:
+            raise RuntimeError('apply_to_event called before resolving frames')
+
+        if self.resolved_frames == NO_CFI_PLACEHOLDER:
+            return False
+
+        self.raw_frames[:] = [frame for module, frame in self.resolved_frames]
+        return True
+
+    @property
+    def needs_cfi(self):
+        """Indicates whether this thread requires reprocessing with CFI due to
+        scanned stack frames."""
+        return any(
+            getattr(FrameTrust, f.get('trust', ''), 0) < MIN_TRUST
+            for f in self.raw_frames
+        )
+
+
+class ThreadProcessingHandle(object):
+    """Helper object for processing all event threads.
+
+    This class offers a view on all threads in the given event payload,
+    including the crashing exception thread. Use ``iter_threads`` to iterate
+    pointers to the original threads' stack traces. Likewise, ``iter_modules``
+    returns references to all modules (images) loaded into the process.
+
+    The handle keeps track of changes to the original data. To signal mutation,
+    call ``indicate_change``. Finally, ``result`` returns the changed data or
+    None if it was not changed.
+    """
+
+    def __init__(self, data):
+        self.data = data
+        self.modules = self._get_modules()
+        self.changed = False
+
+    def _get_modules(self):
+        modules = self.data.get('debug_meta', {}).get('images', [])
+        return ObjectLookup(modules)
+
+    def iter_modules(self):
+        """Returns an iterator over all code modules (images) loaded by the
+        process at the time of the crash. The values are of type ``ObjectRef``.
+        """
+        return self.modules.iter_objects()
+
+    def iter_threads(self):
+        """Returns an iterator over all threads of the process at the time of
+        the crash, including the crashing thread. The values are of type
+        ``ThreadRef``."""
+        for thread in self.data.get('threads', {}).get('values', []):
```

Comment:
Can we leave a comment somewhere that this entire thing will not process exceptions if they are not mentioned in a crashed thread? This is the first time something like this is being done and I'm sure it will be forgotten.

Reply:
Updated. FTR the current implementation would not work at all if the threads are missing. We need thread identifiers to link the threads from the minidump to the event payload.